Allow passing m3u and vfl with other parameters (again)

monkeyman79 opened the pull request:
This reverts commit cd528814507c9fa772c810d8f171776a7f21baa6.
This reverts commit f8be12c0cad25609bb80c9bb35fd355ce3297c16.

- Reading vfl and m3u files specified in 'x64 ...' command line
- Reading vfl and m3u files specified in cmd file
- Executing #COMMAND from m3u file
- Displaying disk image name and possibly disk name on change
  instead of default retroarch message

Fixed since previous attempt:
- Crash if command line was empty
- Crash on insert/eject disk (and/or other operations) if image list was empty
- Deleting temporary file before it was closed

rsn8887 commented:
Can you please adjust your line endings and rebase/force push? All your edits have different line endings and that messes up the diff display and commit, because identical lines where only the endings changed show up as changed.

monkeyman79 commented:
Now it picks correct parameter for autostart and correct image to add to list

rsn8887 commented:
Thanks! I suppose you tested this on Windows only? I can test this on Switch and OSX, unless you already tested it on those platforms? I will try to do that Friday or Saturday, and then I plan to merge.

monkeyman79 commented:
> Thanks! I suppose you tested this on Windows only? I can test this on Switch and OSX, unless you already tested it on those platforms? I will try to do that Friday or Saturday, and then I plan to merge.

Actually I only tested this on Linux.

rsn8887 commented:
I tested on Switch today, and this PR breaks loading content, when content is already loaded, on static platforms.

This needs to be fixed before this can be merged.

Note on static platforms like Switch, Vita, etc., the core is not unloaded when new content is loaded. Therefore, those platforms dependent on retro_load, retro_deinit etc. working correctly. On Desktop platforms, the core is completely removed from memory and reloaded on Load Content, which hides many problems.

rsn8887 commented on libretro/libretro-core.c:
```diff
@@ -2346,15 +2568,15 @@ void retro_run(void)
       /* Load content was called while core was already running, just do a reset with autostart */
       datasette_control(DATASETTE_CONTROL_STOP);
-      autostart_autodetect(RPATH, NULL, 0, AUTOSTART_MODE_RUN);
+      autostart_autodetect(autostartString, NULL, 0, AUTOSTART_MODE_RUN);
```
This functionality if broken in this PR. On static platforms like Switch/Vita, this is used to trigger a reset+autostart when Load Content is selected while the core is running.
 
With this PR, the reset happens but the content is not loaded. I suspect you reset autostartString somewhere.

Note that on shared platforms like Linux, Windows etc., this code is never executed, because the core is unloaded  completely from memory and then reloaded from scratch.

rsn8887 commented on libretro/libretro-core.c:
```diff
@@ -2374,88 +2596,13 @@ void retro_run(void)
-   else
-   {
-      RPATH[0]=0;
+      process_cmdline(info->path);
```
Where are you inserting the disk that was selected (if it was a single d64 file)? You cannot assume pre_main is always executed when retro_load_game is called. Pre_main() is only executed when the core is loaded. On static platforms, that happens only when the core is first started, but not again when Load Content is later selected when the core is already running.

This needs a lot of changes to work correctly again on static platforms.

monkeyman79 replied:
I see the your point, but I don't think it will require much work to address. It just have to behave differently when pre_main has already been called. On the other hand some things won't be possible - the cmd file assumes that you can pass full command line to the x64/xvic/..., which can't be done once pre_main is already called, so the best thing we can to is just read what we can from the cmd - autostart image and/or fliplist and ignore the rest.
I'll take care of that (as soon as I get that bloody wiiu to compile on travis).

sonninnos replied:
> I'll take care of that (as soon as I get that bloody wiiu to compile on travis).

Don't worry about, it always fails. The core build does not fail, but for some reason it also builds RA, which for some other reason does fail.

monkeyman79 replied:
> Don't worry about it, it always fails.

Well, not anymore - see https://github.com/libretro/RetroArch/pull/9496

monkeyman79 commented:
I renamed RPATH to autostartString because the former name didn't tell much and it wasn't clear what it really was

It's simple if we start directly from disk image, like `DISK.D64` - then it is the string that has to be passed as argument to the **main** function, added to image list and passed as argument to autostart after reset. In that case former approach works correctly.

But suppose that you start the game from `ACE2.cmd` with following contents:
`x64 -8 ACE2.d64 -autostart ACE2.vsf`
(this is real example with starts the game immediately without load, but still allows using disk for highscores, loading levels etc.).

Now we have the following:
- Autostart is **ACE2.vsf**
- Disk to attach to and add to fliplist is **ACE2.d64**

Those parameters could be in any order and additional parameters could be present. To get it right we would have to really parse the command line, so duplicate the functionality that vice already does. Instead my approach is to pass the command to vice, let it parse it and ask it what it got out of this.

I wrote earlier
> the cmd file assumes that you can pass full command line to the x64/xvic/..., which can't be done once pre_main is already called

It turns out that actually it's not true - it's possible and quite ok to pass command line arguments after initializing vice, and this is what I do now. I only remove `-fliplist xyz.vfl` or `xyz.m3u` from command line and process it in libretro core.

Now the sequence of loading game is following:
Only when the first game is started:
- Call `stub_main` with the core name as the only argument

Then, when starting first game and any following game:
- Reset vice resources to defaults
- Update vice resources from configuration file
- Update vice resources from libretro environment (opt file and gui)
- Pass command line to vice (new function `initcmdline_restart`)
- Get autostart parameter from vice
- If our fliplist is empty, ask vice for any image attached to tape or drive 8-11 and add it to our fliplist

Bonus is that even if the cmd file contains errors, the vice will still initialize correctly, just reject the cmdline parameters. Previously in case the command line contained error and **skel_main** failed, the core was unusable - it would crash on first call to main loop.

I had one problem with the 'Reset vice resources to defaults' step:
The default value for **(VICII)DoubleSize** parameter is **1**. On startup this sets rendermode to **VIDEO_RENDER_PAL_2X2**, then in `video_canvas_create` we overwrite it to **VIDEO_RENDER_RGB_1X1**, but when later call `resources_set_defaults`, it's set back to PAL2x2 which makes only quarter of screen visible. For now, I disabled DoubleSize resource functionality completely, but I'm sure there's better way to fix that.

I have tested it on Linux, also forcing retro_load_game from emu_reset - it works fine.

Please, review the code and test it on Vita if you can.

monkeyman79 commented:
One more small change incoming

rsn8887 commented:
Ok I still have to test this on Vita/Switch and will report if it works or not.

monkeyman79 commented on libretro/retro_files.c:
```diff
@@ -38,4 +38,16 @@ bool file_exists(const char *filename)
+    size_t dirlen = strlen(basedir);
+    size_t seplen = strlen(RETRO_PATH_SEPARATOR);
+    size_t filelen = strlen(filename);
+    char* result = (char*)malloc(dirlen + seplen + filelen + 1);
```
I would never intentionally ignore such warning :wink:. I must have missed it, or maybe it doesn't show up when compiling for raspberry pi.

rsn8887 commented:
I seems to work fine now in my tests, so I merged. Thanks!

rsn8887 commented:
Ah I had to do a hot fix. Some variables where not inited correctly in reload_restart().

If the user changed a core option, the resource would be changed but the RETROXXX variable would not be changed. If reload_restart() was then called, it would call ui_finalize() and reset the core option back to the RETROXXX variable value.

rsn8887 commented:
Quick question: What is the meaning of START_WITHOUT_PARAMS define and is it necessary?
